File: RAG/src/response_normalizer.py
"""Normalize model output into safe, user-facing RAG response data."""

from __future__ import annotations

import json
import re
from typing import Any


def clean_text(value: Any) -> str:
    if not isinstance(value, str):
        return ""
    text = value.replace("\\r\\n", "\n").replace("\\n", "\n").replace("\r\n", "\n")
    text = re.sub(r"(?is)>{2,}\s*(?:sources?|source)\s*>{2,}.*?(?:<{2,}|$)", "", text)
    text = re.sub(r"(?im)^\s*>{2,}\s*(?:sources?|source)\s*>{2,}\s*$", "", text)
    text = re.sub(r"(?im)^\s*<{2,}\s*$", "", text)
    text = re.sub(r"(?im)^\s*\[?sources?\]?\s*:?\s*$", "", text)
    text = re.sub(r"(?m)^\s*[*]{3,}\s*$", "", text)
    text = re.sub(r"\n{3,}", "\n\n", text)
    return text.strip()


def normalize_response(answer: Any, results: list[dict[str, Any]] | None = None) -> dict[str, Any]:
    if isinstance(answer, dict):
        answer = answer.get("answer", "")
    if isinstance(answer, str):
        try:
            decoded = json.loads(answer)
            if isinstance(decoded, dict):
                answer = decoded.get("answer", answer)
        except json.JSONDecodeError:
            pass
    clean_answer = clean_text(answer)
    sources = []
    for result in results or []:
        metadata = result.get("metadata", {})
        sources.append({
            "title": metadata.get("filename") or metadata.get("title") or result.get("source", "Unknown source"),
            "source_type": metadata.get("source_type", "internal_document"),
            "storage": "minio" if metadata.get("bucket") else None,
            "bucket": metadata.get("bucket"),
            "object_key": metadata.get("object_key"),
            "page": metadata.get("page", result.get("page", 0)),
            "relevance": result.get("relevance_score", result.get("gaussian_score", 0)),
            "published_at": metadata.get("published_at"),
            "url": metadata.get("url"),
        })
    if not clean_answer and not sources:
        return {"answer": "Saya belum menemukan informasi yang cukup relevan untuk menjawab pertanyaan tersebut.", "status": "no_relevant_information", "sources": [], "confidence": None}
    return {"answer": clean_answer, "status": "success", "sources": sources, "confidence": None}

File: frontend/src/lib/responseNormalizer.ts
export type ChatSource = {
  id?: string
  title: string
  domain?: string
  source_type?: string
  storage?: string
  bucket?: string
  object_key?: string
  page?: number
  relevance?: number
  relevance_score?: number
  published_at?: string
  url?: string
}

export type NormalizedChatResponse = {
  answer: string
  status: 'success' | 'partial' | 'no_relevant_information' | 'error'
  sources: ChatSource[]
  confidence: number | null
}

function removeInternalMarkup(value: string): string {
  return value
    .replace(/\\r\\n/g, '\n').replace(/\\n/g, '\n').replace(/\r\n/g, '\n')
    .replace(/>{2,}\s*(?:sources?|source)\s*>{2,}.*?(?:<{2,}|$)/gis, '')
    .replace(/^\s*>{2,}\s*(?:sources?|source)\s*>{2,}\s*$/gim, '')
    .replace(/^\s*<{2,}\s*$/gim, '')
    .replace(/^\s*\[?sources?\]?\s*:?\s*$/gim, '')
    .replace(/^\s*\*{3,}\s*$/gm, '')
    .replace(/\n{3,}/g, '\n\n')
    .trim()
}

export function normalizeChatResponse(payload: any): NormalizedChatResponse {
  let answer = payload?.answer ?? payload?.message ?? ''
  if (typeof answer === 'string') {
    try {
      const decoded = JSON.parse(answer)
      if (decoded && typeof decoded === 'object') answer = decoded.answer ?? answer
    } catch { /* plain text */ }
  }
  answer = removeInternalMarkup(typeof answer === 'string' ? answer : '')
  const sources = Array.isArray(payload?.sources) ? payload.sources.map((source: any) => ({
    title: source.title || source.filename || source.metadata?.filename || source.source || 'Source',
    source_type: source.source_type || source.type || source.metadata?.source_type,
    storage: source.storage || (source.bucket || source.metadata?.bucket ? 'minio' : undefined),
    bucket: source.bucket || source.metadata?.bucket,
    object_key: source.object_key || source.metadata?.object_key,
    page: source.page ?? source.metadata?.page,
    relevance: source.relevance ?? source.relevance_score ?? source.gaussian_score,
    published_at: source.published_at || source.metadata?.published_at,
    url: source.url || source.metadata?.url,
  })) : []
  return {
    answer: answer || (payload?.success === false || payload?.ok === false ? '' : sources.length ? 'Saya menemukan informasi yang relevan.' : ''),
    status: payload?.status === 'error' ? 'error' : sources.length || answer ? payload?.status || 'success' : 'no_relevant_information',
    sources,
    confidence: typeof payload?.confidence === 'number' ? payload.confidence : null,
  }
}
